---
title: "Gut microbiome: Enrichment analysis"
author:
   name: "Dmytro Kryvokhyzha & Mariam Miari"
   email: dmytro.kryvokhyzha@med.lu.se
   affiliation: LUDC Bioinformatics Unit
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    
    
## Which pathways are enriched in the positively/negatively associated species with BMI 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, warning=FALSE)
#knitr::opts_knit$set(root.dir = '~/Science/gut-microbiome/gutmicro_MarjuOH/')
```

## Dependencies

```{r, message=FALSE}
rm(list = ls())
library(fgsea)
library(ggplot2)
library(reshape2)
library(BiocParallel)
library(here)
library(readxl)
library(tidyverse)
library(rio)
```

## Data

### Regression

Test on each file:

  - BMI_full_model

```{r}
reg <- 'BMI_full_model'
associations <- read_excel(here::here("BMI_full_model.xlsx"))
```

Where to save results

```{r}
filePath <- paste('results/', reg, '/', sep = "")
dir.create(filePath, showWarnings = FALSE, recursive=T)
```

Split MGS names and add the association column

```{r}
associations$`-log10(p_value_adj)` <- -log10(associations$p_value_adj)
```

Assign estimates and p-values to MGS:

```{r}
MGSestimate <- associations$est
names(MGSestimate) <- associations$mgs
MGSestimate <- sort(MGSestimate, decreasing = T)

MGSpvalue <- associations$`-log10(p_value_adj)`
names(MGSpvalue) <- associations$MGS
MGSmgspvalue <- sort(MGSpvalue, decreasing = T)
```


```{r eval=FALSE}

mgs2gmm <- readRDS(here::here('upuprf_r1v1_MGS_to_GMM.rds'))
MGS2gmm <- mgs2gmm
```


### GMM

```{bash, eval = F}
wget -v -P intermediate https://raw.githubusercontent.com/raeslab/GMMs/master/GMMs.v1.07.txt
```

Convert GMM database to a table

```{bash, eval = F}
python3 GMMtoTable.py
```

Load the table:

```{r}
gmm <- read.table('intermediate/GMMs.v1.07_table.txt', sep = '\t', header = T)
```

## GMM

### Analysis

```{r}
set.seed(5)
alpha <- 0.05
fgsea_gmm_est <- fgsea(MGS2gmm, MGSestimate,
                      minSize = 1,
                      maxSize = 500,
                      nPermSimple = 100000, # to estimate the p values
                      eps=0)[order(NES, decreasing = T),]
sum(fgsea_gmm_est$padj<=alpha, na.rm = T)
sum(fgsea_gmm_est$padj<=alpha & fgsea_gmm_est$NES>0, na.rm = T)
sum(fgsea_gmm_est$padj<=alpha & fgsea_gmm_est$NES<0, na.rm = T)
```

### Visualization

GSEA

```{r}

fgsea_gmm_est_Up <- fgsea_gmm_est[NES > 0 & padj<=alpha][head(order(pval), n=10), pathway]
fgsea_gmm_est_Down <- fgsea_gmm_est[NES < 0 & padj<=alpha][head(order(pval), n=10), pathway]
fgsea_gmm_est_top <- c(fgsea_gmm_est_Up, rev(fgsea_gmm_est_Down))
plotGseaTable(MGS2gmm[fgsea_gmm_est_top],
              MGSestimate,
              fgsea_gmm_est,
              gseaParam=0.5) #distribution of the pathways across diff MGSs.. if the pathways are more to the left then (+ NES value) then they are upregulated in these MGS. If they are more to the right (- NES), then they are down regulated in these MGS

```

NES + pathway

```{r}
fgsea_gmm_est_plot <- fgsea_gmm_est[fgsea_gmm_est$padj<=alpha,]
fgsea_gmm_est_plot <- fgsea_gmm_est_plot[order(fgsea_gmm_est_plot$size),]
colnames(fgsea_gmm_est_plot)[1] <- 'Module'
fgsea_gmm_est_plot <- merge(fgsea_gmm_est_plot, gmm, by='Module')

pdf(paste(filePath, 'GMM_NES_size.pdf', sep = "/"), width = 9, height = 10)
  ggplot(fgsea_gmm_est_plot, aes(x=NES, y=reorder(name,size), size=size, color=-log(padj))) +
    geom_point() + 
    scale_colour_gradient(low="blue", high="red") +
    labs(y='GMM') +
    geom_vline(xintercept=0, linetype="dotted", size=0.5)
dev.off() # the size of the dot indicates the number of MGS harboring the pathway. Pathways with positive NES means that they are enriched in the positively associated MGS
```

### Annotation

```{r}
fgsea_gmm_est_annot <- merge(as.data.frame(fgsea_gmm_est), as.data.frame(gmm), by=1, all.x = T)
knitr::kable(head(fgsea_gmm_est_annot[which(fgsea_gmm_est_annot$padj<=alpha),
                                     c("pathway", "name", "pval", "padj", "NES")]))
```



## Write results

### Results tables

```{r}
# to convert column of list to vector:
list2vector <- function(df){
  df$MGS <- NA
  for (i in c(1:length(df$leadingEdge))){
    df[i,]$MGS <- paste(unlist(df$leadingEdge[[i]]), collapse=', ')
  }
  df$leadingEdge <- NULL
  return(df)
}

# prepare the results tables for writing
gsea_write_names <- c("pathway", "type", "name", "category",
                      "pval", "padj", "NES", "leadingEdge")


 fgsea_gmm_est_annot_P <- fgsea_gmm_est_annot[,c("pathway", "name",
                                             "pval", "padj", "NES", "leadingEdge", "ko")]
# list columns to vectors:

fgsea_gmm_est_annot_P <- list2vector(fgsea_gmm_est_annot_P)

fgsea_gmm_est_annot_P <- fgsea_gmm_est_annot_P[order(fgsea_gmm_est_annot_P$NES, decreasing = T),]

tax <- read.table("/Volumes/metabogut/Metagenomics_CM/CHAMP/original/files_from_CM/upuprf/MGS/upuprf_r1v1_MGS_taxonomic_annotation.tsv", sep = "\t", header = TRUE)

maintaxa <- tax %>% 
  unite("maintaxa", species, MGS, sep = ".", remove = FALSE)


replace_values <- function(df, df2) {
  df$mgs <- strsplit(df$MGS, ", ", fixed = TRUE)
  
  
  # Iterate over each row
  df$MGS <- sapply(df$mgs, function(x) {
    # match values in col2 with MGS from maintaxa and replace them with corresponding maintaxa values
    replacements <- ifelse(x %in% maintaxa$MGS, maintaxa$maintaxa[match(x, maintaxa$MGS)], x)
    paste(replacements, collapse = ", ")
  })
  
  return(df)
}
result <- replace_values(fgsea_gmm_est_annot_P, maintaxa)

result <- result[,c(1:ncol(result)-1)]
```

```{r}
export(result,
            paste(filePath, 'enrichment_fgsea_GMM.xlsx', sep='/'))
```

## how many unique species are there for those with significant positive NES
```{r}
res <- read.csv("/Volumes/research/metabogut/BMI_2021/analysisv3.0/enrichment/results/BMI_full_model/enrichment_fgsea_GMM.csv", header = T, sep = "\t")


res_sig_pos <- res %>% filter (NES > 0 & padj < 0.05)
res_sig_pos_separated <- res_sig_pos %>%
  separate_rows(MGS, sep = "\\|") 

length(unique(res_sig_pos_separated$MGS))



res_sig_neg <- res %>% filter (NES < 0 & padj < 0.05)
res_sig_neg_separated <- res_sig_neg %>%
  separate_rows(MGS, sep = "\\|") 

length(unique(res_sig_neg_separated$MGS))


```
