---
title: "Top100MGS_heatmap"
output: html_document
date: "2023-02-28" 
"This script plots the top 100 significant MGS in BMI, WaistadjBMI, WHRadjBMI all together in one plot and separated. The significant associations are for all phenotypes and the estimates of these associations are plotted even if significant in one phenotype but not the other."
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Run packages
```{r}
rm(list = ls())

library(readxl)
library(tidyverse)
library(dbplyr)
library(dplyr)
library(ggplot2)
library(stringr)
library(ggtext)
library(reshape2)
library(rio)
library(xtable)
library(gridExtra)
library(gtable)
library(pheatmap)
library(ComplexHeatmap)
library(circlize)
library(grid)
library(tidyr)
library(here)
```


### read data
```{r}
load("/Volumes/research/metabogut/BMI_2021/marlena/results_version_221109/all_results_221109.Rdata")

taxonomy <- read_xlsx("/Volumes/research/metabogut/Metagenomics_CM/Data LUNGUT/Updated_taxonomy_annotation_AronEklundCM211021/HG3.A.7_tax_updatedTaxonomy_NCBIversion210502_fromAronEklundCM211013.xlsx")
```

### edit MGS names
```{r}
taxonomy <- taxonomy %>% rename(MGS = ...1)
maintaxa <- taxonomy %>% unite("maintaxa", MainTax:MGS, sep = " ", remove = FALSE)
maintaxa <- maintaxa %>% dplyr::select(maintaxa, MGS)
maintaxa <- maintaxa %>% rename(mgs = MGS)

#these results do not have the interaction term (e.g. MGS:maleTrue is ignored)
#annot = annotated MGS
all_results_annot <- merge(all_results, maintaxa, by = 'mgs') #988 unique MGS

all_results_annot <- all_results_annot %>% 
    separate(maintaxa, into = c("genus", "species", "other"), 
             sep = "\\s(?=sp\\.|subsp\\.)|\\s+(?=HG3A)", 
             extra = "merge", remove = FALSE) %>% 
    unite(others, species, other, na.rm = TRUE, sep = " ")

all_results_annot$label=paste("*",all_results_annot$genus,"* ",all_results_annot$other,sep="")
```

### take all_cohorts, main analysis, full models for each of the 3 phenotypes
###### we have more significant associations with WHR than waist, correct?
```{r}
#filter p_fdr_per_outcome in the whole data first
sig_res_mgs <- subset(all_results_annot, p_fdr_per_outcome <= 0.05)
#BMI
sig_res_mgs_all_main_full_BMI <- sig_res_mgs[which(sig_res_mgs$model_name == 'full_M1_BMI' & sig_res_mgs$cohort == 'all_cohorts' & sig_res_mgs$analysis == 'main'),] #371 sig associations
#export(sig_res_mgs_all_main_full_BMI, '/Volumes/research/metabogut/BMI_2021/sig_res_BMI.xlsx')

#waist
sig_res_mgs_all_main_full_waist <- sig_res_mgs[which(sig_res_mgs$model_name == 'full_M3_waist_adjBMI' & sig_res_mgs$cohort == 'all_cohorts' & sig_res_mgs$analysis == 'main'),] #31 sig associations
#export(sig_res_mgs_all_main_full_waist, '/Volumes/research/metabogut/BMI_2021/sig_res_waist.xlsx')

#WHR
sig_res_mgs_all_main_full_whr <- sig_res_mgs[which(sig_res_mgs$model_name == 'full_M2_WHR_adjBMI' & sig_res_mgs$cohort == 'all_cohorts' & sig_res_mgs$analysis == 'main'),] #52 sig associations
#export(sig_res_mgs_all_main_full_whr, '/Volumes/research/metabogut/BMI_2021/sig_res_whr.xlsx')

all_sig_pheno_all_main_full <- rbind(sig_res_mgs_all_main_full_BMI,sig_res_mgs_all_main_full_waist, sig_res_mgs_all_main_full_whr) #454 significant associations across all phenotypes

all_sig_pheno_all_main_full <- all_sig_pheno_all_main_full %>% 
  mutate(phenotype = ifelse(model_name == 'full_M1_BMI', 'BMI',
                            ifelse(model_name == 'full_M3_waist_adjBMI','Waist_adjBMI','WHR_adj_BMI'))) 

#save results
#export(all_sig_pheno_all_main_full, file = here('results/output_files/gmm', 'all_sig_res_annotated.xlsx'), overwrite = TRUE)

```

### take top 100 associations across all phenotypes
```{r}
all_sig_pheno_all_main_full_100 <- all_sig_pheno_all_main_full %>% 
  filter(rank(p_fdr_per_outcome)<=100)
```


### add estimates of MGSs that are significant in one phenotype but not the other
```{r}
#BMI
res_mgs_all_main_full_BMI <- all_results[which(all_results$model_name == 'full_M1_BMI' & all_results$cohort == 'all_cohorts' & all_results$analysis == 'main'),]

#waist
res_mgs_all_main_full_waist <- all_results[which(all_results$model_name == 'full_M3_waist_adjBMI' & all_results$cohort == 'all_cohorts' & all_results$analysis == 'main'),]

#WHR
res_mgs_all_main_full_whr <- all_results[which(all_results$model_name == 'full_M2_WHR_adjBMI' & all_results$cohort == 'all_cohorts' & all_results$analysis == 'main'),]


#add them all on top of each other
all_pheno_all_main_full <- rbind(res_mgs_all_main_full_BMI,res_mgs_all_main_full_waist, res_mgs_all_main_full_whr)

all_pheno_all_main_full <- all_pheno_all_main_full %>% 
  mutate(phenotype = ifelse(model_name == 'full_M1_BMI', 'BMI',                                                                 ifelse(model_name == 'full_M3_waist_adjBMI', 'Waist_adjBMI','WHR_adj_BMI')))
```

### merge with all_sig_pheno_all_main_full to get the estimates of MGS that are for instance significant in BMI but not in whradjbmi
```{r}
merge_all_with_sig <- merge(all_sig_pheno_all_main_full_100, all_pheno_all_main_full , by = 'mgs', all = FALSE)
```

### subset data
```{r}
merge_all_with_sig_subset <- merge_all_with_sig %>% 
  transmute(mgs = mgs, maintaxa = maintaxa, analysis = analysis.y, 
            outcome = outcome.y,
            cohort = cohort.y, model_name = model_name.y, 
            n_obs = n_obs.y, est = est.y,
            se = se.y,ci_low = ci_low.y, ci_high = ci_high.y, p = p.y, 
            p_fdr_per_outcome = p_fdr_per_outcome.y,
            phenotype = phenotype.y,
            label = label)

merge_all_with_sig_subset <- merge_all_with_sig_subset %>% 
  mutate(phenotype = ifelse(phenotype == 'WHR_adj_BMI', 'WHRadjBMI', 
                            ifelse(phenotype == 'Waist_adjBMI', 'WaistadjBMI', 'BMI')))

```

### remove duplicated values 
```{r}
merge_all_with_sig_subset <- unique(merge_all_with_sig_subset)
```

```{r}
to_table <- merge_all_with_sig_subset[,c(1,2,8,13,14,15)]
```

### use spread from tidyr to transform column values to colnames
```{r}
transform_to_table <- to_table %>% spread(phenotype, p_fdr_per_outcome)
transform_to_table<- transform_to_table %>%
  group_by(mgs) %>%
  summarise( maintaxa = label,
             BMI = na.omit(BMI),
             WaistadjBMI = na.omit(WaistadjBMI),
             WHRadjBMI = na.omit(WHRadjBMI))

transform_to_table <- unique(transform_to_table)

transform_to_table <- transform_to_table %>% 
  mutate(BMI = signif(BMI,2),
         WaistadjBMI = signif(WaistadjBMI,2),
         WHRadjBMI = signif(WHRadjBMI,2))
```

### plot all phenotypes together
```{r}
P <- merge_all_with_sig_subset %>%  group_by (phenotype) %>% ggplot(aes(x=reorder(factor(label),est), y=est, ymin=ci_low, ymax=ci_high)) +
  geom_pointrange(position = position_dodge(width = 1), size=0.1,aes(color=phenotype), shape = 20) + 
  scale_color_manual(values = c("grey30", "mediumorchid", "gold2"), labels = c("BMI", "WaistadjBMI", "WHRadjBMI")) + 
  xlab("") + 
  ylab('est (95% CI)') + 
  ggtitle("MGS") + 
 # coord_cartesian(ylim = c(-5, 30))+ 
  coord_flip() + theme_minimal() + #scale_y_reverse() +
  theme(axis.text.y = element_markdown(size= 6.5),
        axis.text.x = element_markdown(size = 6.5),
        axis.title.x = element_text(size = 9),
        plot.title = element_text(size = 9.2,hjust = -0.09),
        legend.position = 'none',
        #legend.position = c(0.9,1.5),
        #legend.position = 'bottom',
       # legend.direction = 'horizontal',
        plot.margin=grid::unit(c(2,0,0,0), "mm")) 
P
```

```{r}
#get y axis labels in the order they were plotted
maintaxa_in_P <- ggplot_build(P)$layout$panel_params[[1]]$y$get_labels()
#flip bottom to top
maintaxa_in_P<- rev(maintaxa_in_P)
```

### before plotting the heatmap, make sure that the order of the table maintaxa matches the ones in the plot (P)
```{r}
# Extract the levels of the text_column  in df1 in the same order as the unique values in df2
levels_df1 <- levels(as.factor(transform_to_table$maintaxa))[match(maintaxa_in_P, levels(as.factor(transform_to_table$maintaxa)))]

# Find the indices of the key values in the text_column column of the data frame
indices <- match(levels_df1, transform_to_table$maintaxa)

# Reorder the values in the text_column column based on the indices
transform_to_table <- transform_to_table[indices,]
```

### start plotting each pehnotype separately
### BMI
```{r}
#take BMI data
data_bmi_plot <- merge_all_with_sig_subset[merge_all_with_sig_subset$phenotype == 'BMI',]

P_bmi <- data_bmi_plot %>% ggplot(aes(x=reorder(factor(label),est), y=est, ymin=ci_low, ymax=ci_high)) +
  geom_pointrange(size=0.1,aes(color=phenotype), shape = 20) + 
  scale_color_manual(values = 'grey30', labels = 'BMI') + 
  xlab("") + 
  ylab('est (95% CI)') + 
  ggtitle("MGS") + 
  # coord_cartesian(ylim = c(-5, 30))+ 
  coord_flip() + theme_minimal() + 
  geom_hline(yintercept = 0, color = "grey80")+
  theme(axis.text.y = element_markdown(size= 5),
        axis.text.x = element_markdown(size = 6.5),
        axis.title.x = element_text(size = 9),
        plot.title = element_text(size = 9.2,hjust = -0.3),
        legend.position = 'none',
        #legend.position = c(0.9,1.5),
        #legend.position = 'bottom',
        # legend.direction = 'horizontal',
        plot.margin=grid::unit(c(2,0,0,0), "mm")) 
P_bmi

#get y axis labels in the order they were plotted
maintaxa_in_P_bmi <- ggplot_build(P_bmi)$layout$panel_params[[1]]$y$get_labels()
#flip bottom to top
maintaxa_in_P_bmi<- rev(maintaxa_in_P_bmi)


###### before plotting the heatmap, make sure that the order of the table Description matches the ones in the plot (P)

# Extract the levels of the text_column  in df1 in the same order as the unique values in df2
levels_df1_bmi <- levels(as.factor(transform_to_table$maintaxa))[match(maintaxa_in_P_bmi, levels(as.factor(transform_to_table$maintaxa)))]

# Find the indices of the key values in the text_column column of the data frame
indices_bmi <- match(levels_df1_bmi, transform_to_table$maintaxa)

# Reorder the values in the text_column column based on the indices
transform_to_table <- transform_to_table[indices_bmi,]

```
### heatmap: BMI
```{r}
data_bmi_hp <- transform_to_table[,3]
data_bmi_hp[nrow(data_bmi_hp) +1, ] <- 0 
data_bmi_hp[nrow(data_bmi_hp) +1, ] <- 1 
data_bmi_hp <- as.matrix(data_bmi_hp)
#colnames(data) <- NULL

#create this foo variable just to add numbers that are not numeric to the table because some numeric numbers are being converted to e notation and others are not
#foo_bmi_hp <- as.data.frame(transform_to_table %>% 
                     #  mutate(BMI = formatC(BMI, format = 'e', digits = 2)))
#names(foo) <- NULL

#foo_bmi_hp[nrow(foo_bmi_hp) +1, ] <- 0 
#foo_bmi_hp[nrow(foo_bmi_hp) +1, ] <- 1 

#setting colors
clrsp <- colorRampPalette(c("red", "white"))   
clrs <- clrsp(nrow(data_bmi_hp)) 
#set breaks from 0 to 0.05
breaks1 <- seq(-log10(0+1e-50), -log10(1), length.out = nrow(data_bmi_hp))

data_bmi_hp[100,1] <- data_bmi_hp[100,1] + 1e-50


p_bmi_hp <- grid.grabExpr(draw(ComplexHeatmap::pheatmap(-log10(data_bmi_hp),
                                                        cluster_rows = F,
                                                        cluster_cols = F, 
                                                        cellwidth=42, 
                                                        cellheight = 8.2,
                                                        display_numbers =
                                                        data_bmi_hp,
                                                        column_names_side =
                                                          c("top"),name = "p
                                                        value",
                                                        angle_col = c("0"), 
                                                        number_color = "black", 
                                                        border_color = NA,
                                                        fontsize_col = 9.3,
                                                        fontsize = 8.2,
                                                        legend = FALSE,
                                                        col =clrs, breaks =
                                                          breaks1)))
all_hp_bmi <- grid.arrange( P_bmi,p_bmi_hp, widths = c(2, 1))

```

### Waist
```{r}
data_waist_plot <- merge_all_with_sig_subset[merge_all_with_sig_subset$phenotype == 'WaistadjBMI',]

# First match the maintaxa in waist with that in bmi (because all plots should be aligned with the same descriptions to the left)
levels_df1_waist <- levels(as.factor(data_waist_plot$label))[match(maintaxa_in_P_bmi, levels(as.factor(data_waist_plot$label)))]

# Find the indices of the key values in the text_column column of the data frame to fix the heatmap
#I am using transform_to_table dataframe because it contains the exponentiated pvalues which are to be added in heatmap cells
indices_waist_hp <- match(levels_df1_waist, transform_to_table$maintaxa)

# Reorder the values in the text_column column based on the indices
transform_to_table_waist_hp <- transform_to_table[indices_waist_hp,] #this is for the heatmap

#----- #the task here is to match the description in waist dataframe to that in BMI so that when they are plotted in same plot, the descriptions are aligned in both phenotypes

# pr = this is used for point range plot
indices_waist_pr <- match(levels_df1_waist, data_waist_plot$label)

# Reorder the values in the text_column column based on the indices
data_waist_plot <- data_waist_plot[indices_waist_pr,]

#this is to force plot the y axis labels the order they are in the dataframe
data_waist_plot$label <- factor(data_waist_plot$label, levels=data_waist_plot$label)



P_waist <- data_waist_plot %>% ggplot(aes(x=maintaxa, y=est, ymin=ci_low, ymax=ci_high)) +
  geom_pointrange(size=0.1,aes(color=phenotype), shape = 20) + 
  scale_color_manual(values = 'mediumorchid', labels = 'WaistadjBMI') + scale_x_discrete(limits=rev)+
  xlab("") + 
  ylab('est (95% CI)') + 
  ggtitle("") + 
  # coord_cartesian(ylim = c(-5, 30))+ 
  coord_flip() + theme_minimal() + 
  geom_hline(yintercept = 0, color = "grey80")+
  theme(#axis.text.y = element_text(size= 6.5),
        axis.text.x = element_text(size = 6.5),
        axis.title.x = element_text(size = 9),
        plot.title = element_text(size = 9.2,hjust = -0.09),
        #plot.title = element_blank(),
        axis.ticks.y=element_blank(),
        axis.text.y=element_blank(),
        legend.position = 'none',
        #legend.position = c(0.9,1.5),
        #legend.position = 'bottom',
        # legend.direction = 'horizontal',
        plot.margin=grid::unit(c(2,0,0,0), "mm")) 
P_waist
```
### Heatmap: Waist
```{r}
data_waist_hp <- transform_to_table_waist_hp[,4]
data_waist_hp[nrow(data_waist_hp) +1, ] <- 0 
data_waist_hp[nrow(data_waist_hp) +1, ] <- 1 


data_waist_hp <- as.matrix(data_waist_hp)
#colnames(data) <- NULL

#create this foo variable just to add numbers that are not numeric to the table because some numeric numbers are being converted to e notation and others are not
#foo_waist_hp <- as.data.frame(transform_to_table %>% 
                             # mutate(WaistadjBMI = formatC(WaistadjBMI, format = 'e', digits #= 2)))
#names(foo) <- NULL

#foo_waist_hp[nrow(foo_waist_hp) +1, ] <- 0 
#foo_waist_hp[nrow(foo_waist_hp) +1, ] <- 1 

data_waist_hp[100,1] <- data_waist_hp[100,1] + 1e-50

p_waist_hp <- grid.grabExpr(draw(ComplexHeatmap::pheatmap(-log10(data_waist_hp),
                                                          cluster_rows = F,
                                                          cluster_cols = F, 
                                                          cellwidth=42,
                                                          cellheight= 8.2,
                                                          display_numbers = data_waist_hp, 
                                                          column_names_side =
                                                          c("top"),
                                                          name = "pvalue",
                                                          angle_col = c("0"),
                                                          number_color = "black", 
                                                          border_color = NA,
                                                          fontsize_col = 9.3,
                                                          fontsize = 8.2,
                                                          legend = FALSE,
                                                          col =clrs, 
                                                          breaks = breaks1)))

grid.arrange( P_bmi,p_bmi_hp, P_waist, p_waist_hp, widths = c(2,1,1,1))
```

### WHR
```{r}
data_whr_plot <- merge_all_with_sig_subset[merge_all_with_sig_subset$phenotype == 'WHRadjBMI',]

# First match the description in whr with that in bmi (because all plots should be aligned with the same descriptions to the left)
levels_df1_whr <- levels(as.factor(data_whr_plot$label))[match(maintaxa_in_P_bmi, levels(as.factor(data_whr_plot$label)))]

indices_whr_hp <- match(levels_df1_whr, transform_to_table$maintaxa)

# Reorder the values in the text_column column based on the indices
transform_to_table_whr_hp <- transform_to_table[indices_whr_hp,] #this is for the heatmap

#----- #the task here is to match the description in whr dataframe to that in BMI so that when they are plotted in same plot, the descriptions are aligned in both phenotypes

# pr = this is used for point range plot
indices_whr_pr <- match(levels_df1_whr, data_whr_plot$label)

# Reorder the values in the text_column column based on the indices
data_whr_plot <- data_whr_plot[indices_whr_pr,]

data_whr_plot$label <- factor(data_whr_plot$label, levels=data_whr_plot$label)


P_whr <- data_whr_plot %>% ggplot(aes(x=label, y=est, ymin=ci_low, ymax=ci_high)) +
  geom_pointrange(size=0.1,aes(color=phenotype), shape = 20) + 
  scale_color_manual(values = 'orange', labels = 'WHRadjBMI') + scale_x_discrete(limits=rev)+
  xlab("") + 
  ylab('est (95% CI)') + 
  ggtitle("") + 
  # coord_cartesian(ylim = c(-5, 30))+ 
  coord_flip() + theme_minimal() + 
  geom_hline(yintercept = 0, color = "grey80")+#scale_y_reverse() +
  theme(#axis.text.y = element_text(size= 6.5),
    axis.text.x = element_text(size = 6.5),
    axis.title.x = element_text(size = 9),
    plot.title = element_text(size = 9.2,hjust = -0.09),
    #plot.title = element_blank(),
    axis.ticks.y=element_blank(),
    axis.text.y=element_blank(),
    legend.position = 'none',
    #legend.position = c(0.9,1.5),
    #legend.position = 'bottom',
    # legend.direction = 'horizontal',
    plot.margin=grid::unit(c(2,0,0,0), "mm")) 
P_whr


```
### Heatmap: WHR
```{r}
data_whr_hp <- transform_to_table_whr_hp[,5]
data_whr_hp[nrow(data_whr_hp) +1, ] <- 0 
data_whr_hp[nrow(data_whr_hp) +1, ] <- 1 

data_whr_hp <- as.matrix(data_whr_hp)
#colnames(data) <- NULL

#create this foo variable just to add numbers that are not numeric to the table because some numeric numbers are being converted to e notation and others are not
#foo_whr_hp <- as.data.frame(transform_to_table %>% 
 #                               mutate(WHRadjBMI = formatC(WHRadjBMI, 
  #                                                         format = 'e', 
   #                                                        digits = 2)))
#names(foo) <- NULL

#foo_whr_hp[nrow(foo_whr_hp) +1, ] <- 0 
#foo_whr_hp[nrow(foo_whr_hp) +1, ] <- 1 

data_whr_hp[100,1] <- data_whr_hp[100,1] + 1e-50

p_whr_hp <- grid.grabExpr(draw(ComplexHeatmap::pheatmap(-log10(data_whr_hp),
                                                        cluster_rows = F,
                                                        cluster_cols = F, 
                                                        cellwidth=42, 
                                                        cellheight = 8.2,
                                                        display_numbers =
                                                        data_whr_hp,
                                                        column_names_side =
                                                          c("top"),
                                                        name = "pvalue",
                                                        angle_col = c("0"), 
                                                        number_color = "black", 
                                                        fontsize_col = 9.3,
                                                        fontsize = 8.2,
                                                        border_color = NA,
                                                        col =clrs, 
                                                        breaks = breaks1)))

```

### Final format for everything together
```{r}
#add white rectangle
rect_whr <- rectGrob(gp=gpar(fill='white', col = 'white'), x=unit(-1.4,"npc"),
                     y=unit(0.003,"npc"),
                     width=unit(1.3,"npc"), height=unit(0.048,"npc"))

rect_waist <- rectGrob(gp=gpar(fill='white', col = 'white'), x=unit(-8,"npc"),
                       y=unit(0.003,"npc"),
                      width=unit(1.5,"npc"), height=unit(0.048,"npc"))

rect_bmi <- rectGrob(gp=gpar(fill='white', col = 'white'), x=unit(-15,"npc"),
                     y=unit(0.003,"npc"),
                       width=unit(1.5,"npc"), height=unit(0.048,"npc"))


all_plots <- grid.arrange( P_bmi,p_bmi_hp,P_waist, p_waist_hp, P_whr, 
                           p_whr_hp,rect_whr, rect_waist, rect_bmi, widths =
                             c(3.5,1,2,1,2,1,0.5,0.5,0.5))

ggsave('/Users/ma8244mi/Desktop/BMI-paper/plots/Top100MGS_heatmap.pdf', all_plots, width = 50,height = 30, units = 'cm')

```

