---
title: "waist_whr_AFMI_plots"
output: html_document
date: "2023-06-10"
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = '/Volumes/research/metabogut/BMI_2021')
```

Load data from Marlenas file 
```{r}
rm(list = ls())
library(readxl)
library(tidyverse)
library(gridExtra)
library(ggplot2)
library(RColorBrewer)
library(pheatmap)
library(ComplexHeatmap)
library(rio)

load("marlena/results_version_221109/all_results_221109.Rdata")
load("Louise_intermediate/data_log_filt5_n12028_update220608_221104.RData")
taxonomy <- read_xlsx("/Volumes/research/metabogut/Metagenomics_CM/Data LUNGUT/Updated_taxonomy_annotation_AronEklundCM211021/HG3.A.7_tax_updatedTaxonomy_NCBIversion210502_fromAronEklundCM211013.xlsx")

```

##################################################################
##################################################################
##################################################################
##################################################################
##################################################################
###### start with waist

Add in taxonomical name 
```{r}
taxonomy <- taxonomy %>% rename(MGS = ...1)
maintaxa <- taxonomy %>% unite("maintaxa", MainTax:MGS, sep = " ", remove = FALSE)
maintaxa <- maintaxa %>% dplyr::select(maintaxa, MGS)
maintaxa <- maintaxa %>% rename(mgs = MGS)

#remove the :maleTRUE inorder to merge with maintaxa and get annotations
all_results$mgs_only <- sub(":maleTRUE", "", all_results$mgs)

# 988 mgs 
data_all <- merge(maintaxa, all_results, by.x = "mgs", by.y = "mgs_only")
colnames(data_all)[1] <- 'mgs_only'
colnames(data_all)[names(data_all) == 'mgs.y'] <- 'mgs'


# Split MGS names into genus-species and other columns
data_sep <- data_all %>% 
    separate(maintaxa, into = c("genus", "species", "other"), 
             sep = "\\s(?=sp\\.|subsp\\.)|\\s+(?=HG3A)", 
             extra = "merge", remove = FALSE) %>% 
    unite(others, species, other, na.rm = TRUE, sep = " ")

data_sep$label=paste("*",data_sep$genus,"* ",data_sep$other,sep="")
```

Extract significant waist results 
```{r}
waist_sig <- data_sep %>% filter(outcome == "waist", analysis == "main", 
                               model_name == "full_M3_waist_adjBMI",
                               cohort == "all_cohorts",p_fdr_per_outcome <= 0.05)

#take top 100 most significant associations
top_waist_sig <- waist_sig %>% filter(p_fdr_per_outcome<=0.05)
#remove mgs_only column
top_waist_sig <- top_waist_sig[2:ncol(top_waist_sig)]
```

Add sex specific results 

full_M3_waist_adjBMI
full_stratM3_waist_adjBMI_females
full_stratM3_waist_adjBMI_males
```{r}
waist_sex <- data_sep %>% filter(analysis == "main") %>% filter(
    model_name == "full_stratM3_waist_adjBMI_females" | 
        model_name =="full_stratM3_waist_adjBMI_males", cohort == 'all_cohorts')

sig_mgs_list <- top_waist_sig %>% dplyr::select(mgs)

waist_sex_top <- merge(waist_sex, sig_mgs_list, by = 'mgs')
waist_sex_top <- waist_sex_top[,c(3:9,1, 10:ncol(waist_sex_top))]
```

Add interaction results
```{r}
subset_data_sep <- data_sep[grepl(":maleTRUE", data_sep$mgs),]


waist_int <- subset_data_sep %>% filter(analysis == "main",
                               cohort == 'all_cohorts', model_name == "full_intM3_waist_adjBMI" ) 

waist_int_top <- merge(waist_int, sig_mgs_list, by.x = 'mgs_only', by.y = 'mgs')

#keep mgs_only column and remove the one with :maleTRUE because I will merge with mgs_only below
waist_int_top <- waist_int_top[c(2:8,1,10:ncol(waist_int_top))]
names(waist_int_top)[names(waist_int_top) == "mgs_only"] <- "mgs"
```

### put data data together
```{r}

merge_all_with_sig <- rbind(top_waist_sig, waist_sex_top,waist_int_top)

# subset the needed data
merge_all_with_sig_subset <- merge_all_with_sig %>% 
  transmute(mgs = mgs, maintaxa = maintaxa, label = label,
            analysis = analysis, outcome = outcome,
            cohort = cohort, model_name = model_name, 
            n_obs = n_obs, est = est,
            se = se,ci_low = ci_low, ci_high = ci_high, p = p, 
            p_fdr_per_outcome = p_fdr_per_outcome)


merge_all_with_sig_subset <- merge_all_with_sig_subset %>% 
  mutate(phenotype = ifelse(model_name == 'full_stratM3_waist_adjBMI_females', 'Female', 
                            ifelse(model_name == 'full_stratM3_waist_adjBMI_males', 'Male', 
                                   ifelse(model_name == 'full_M3_waist_adjBMI', 'All',
                                          'Interaction'))))
```

### re-structure the data to plot in the heatmap
```{r chunk created by Mariam 10 feb 2023}
transform_to_table <- merge_all_with_sig_subset %>% spread(phenotype, p_fdr_per_outcome)

transform_to_table<- transform_to_table %>%
  group_by(mgs) %>%
  summarise( label = label,
             All = na.omit(All),
             Female = na.omit(Female),
             Male = na.omit(Male),
             Interaction = na.omit(Interaction))


transform_to_table <- unique(transform_to_table) 

#use this table for the heatmap
transform_to_table <- transform_to_table %>% 
  mutate(All = signif(All,2),
         Female = signif(Female,2),
         Male = signif(Male,2),
         Interaction = signif(Interaction,2))

```


```{r}

#remove interaction estimates
data_plot <- merge_all_with_sig_subset[merge_all_with_sig_subset$phenotype != 'Interaction',]
P <- data_plot %>%  group_by (phenotype) %>% ggplot(aes(x=reorder(factor(label),est), y=est, ymin=ci_low, ymax=ci_high)) +
    geom_hline(yintercept = 0, color = "grey80")+
  geom_pointrange(position = position_dodge(width = 1), size=0.1,aes(color=phenotype), shape = 20) + 
  scale_color_manual(values = c("black", "hotpink", "green4"), labels = c("All", "Female", "Male")) + 
  xlab("") + 
  ylab('') + 
  #ggtitle("maintaxa") + 
 # coord_cartesian(ylim = c(-5, 30))+ 
  coord_flip() + theme_minimal() + #scale_y_reverse() +
  theme(axis.text.y = element_blank(),
        axis.text.x = element_text(size = 6.5),
        #axis.title.x = element_text(size = 9),
        #plot.title = element_text(size = 9.2,hjust = -0.09),
        legend.position = 'none',
        plot.margin=grid::unit(c(2,0,-1,0), "mm")) 
        #legend.position = c(0.9,1.5),
        #legend.position = 'bottom',
       # legend.direction = 'horizontal',
        
P


```

### before plotting the heatmap make sure that the order of maintaxa matches order of estimates
```{r}
#get y axis labels in the order they were plotted
desc_in_P <- ggplot_build(P)$layout$panel_params[[1]]$y$get_labels()
#flip bottom to top
desc_in_P<- rev(desc_in_P)


# Extract the levels of the text_column  in df1 in the same order as the unique values in df2
levels_df1 <- levels(as.factor(transform_to_table$label))[match(desc_in_P, levels(as.factor(transform_to_table$label)))]

# Find the indices of the key values in the text_column column of the data frame
indices <- match(levels_df1, transform_to_table$label)

# Reorder the values in the text_column column based on the indices
transform_to_table <- transform_to_table[indices,]
```

#plot both plots together
```{r}
data <- transform_to_table[,c(3:6)]
data[nrow(data) +1, ] <- 0 + 1e-50
data[nrow(data) +1, ] <- 1 

#colnames(data) <- NULL

#create this foo variable just to add numbers that are not numeric to the table because some numeric numbers are being converted to e notation and others are not
#foo <- as.data.frame(transform_to_table %>% 
 # mutate(All = formatC(All, format = 'e', digits = 2),
  #       Female = formatC(Female,format = 'e', digits = 2),
   #      Male = formatC(Male,format = 'e', digits = 2)))
#names(foo) <- NULL

#foo[nrow(foo) +1, ] <- 0
#foo[nrow(foo) +1, ] <- 1 

#setting colors
clrsp <- colorRampPalette(c("red", "white"))   
clrs <- clrsp(nrow(data)) 

#set breaks 
breaks1 <- seq(-log10(0+1e-50), -log10(1), length.out = nrow(data))

p <- grid.grabExpr(draw(ComplexHeatmap::pheatmap(as.matrix(-log10(data)),cluster_rows = F, 
                                                 cluster_cols = F, cellwidth=50,
                                                 show_colnames = F, fontsize = 7.1,
                                                 display_numbers = as.matrix(data), 
                                                 column_names_side = c("top"),name ="pvalue",
                                                 angle_col = c("0"),  
                                                 number_color =  "black", 
                                                 gaps_col =c(1,2,3),
                                                 col =clrs, breaks = breaks1)))

#draw a white rectangle on the top of the plotted NAs
rect <- rectGrob(gp=gpar(fill='white', col = 'white'), x=unit(-2.5,"npc"),
                 y=unit(0.0035,"npc"),
                 width=unit(9.2,"npc"), height=unit(0.048,"npc"))


#ggsave('/Users/ma8244mi/Desktop/MGS-heatmap.pdf', all_plots, width = 30,height = 40, units = 'cm')

```


```{r}
merge_all_with_sig_subset <- merge_all_with_sig_subset[merge_all_with_sig_subset$phenotype != 'Interaction',]
merge_all_with_sig_subset <- merge_all_with_sig_subset[order(merge_all_with_sig_subset$mgs),]
export(merge_all_with_sig_subset, '/Volumes/research/metabogut/BMI_2021/summaryresults/MGS/waist_sigAll_females_males.xlsx')
```

#####################################################################
#####################################################################
#####################################################################
#####################################################################

###### start with whr

```{r}
whr_sig <- data_sep %>% filter(outcome == "WHR", analysis == "main", 
                               model_name == "full_M2_WHR_adjBMI",
                               cohort == "all_cohorts",p_fdr_per_outcome <= 0.05)

#take top 100 most significant associations
top_whr_sig <- whr_sig %>% filter(p_fdr_per_outcome<=0.05)
#remove mgs_only column
top_whr_sig <- top_whr_sig[2:ncol(top_whr_sig)]
```

Add sex specific results 

full_M2_WHR_adjBMI
full_stratM2_WHR_adjBMI_females
full_stratM2_WHR_adjBMI_males
```{r}
whr_sex <- data_sep %>% filter(analysis == "main") %>% filter(
    model_name == "full_stratM2_WHR_adjBMI_females" | 
        model_name =="full_stratM2_WHR_adjBMI_males", cohort == 'all_cohorts')

sig_mgs_list <- top_whr_sig %>% dplyr::select(mgs)

whr_sex_top <- merge(whr_sex, sig_mgs_list, by = 'mgs')
whr_sex_top <- whr_sex_top[,c(3:9,1, 10:ncol(whr_sex_top))]
```

Add interaction results
```{r}
subset_data_sep <- data_sep[grepl(":maleTRUE", data_sep$mgs),]


whr_int <- subset_data_sep %>% filter(analysis == "main",
                               cohort == 'all_cohorts', model_name == "full_intM2_WHR_adjBMI" ) 

whr_int_top <- merge(whr_int, sig_mgs_list, by.x = 'mgs_only', by.y = 'mgs')

#keep mgs_only column and remove the one with :maleTRUE because I will merge with mgs_only below
whr_int_top <- whr_int_top[c(2:8,1,10:ncol(whr_int_top))]
names(whr_int_top)[names(whr_int_top) == "mgs_only"] <- "mgs"
```

### put data data together
```{r}

merge_all_with_sig <- rbind(top_whr_sig, whr_sex_top,whr_int_top)

# subset the needed data
merge_all_with_sig_subset <- merge_all_with_sig %>% 
  transmute(mgs = mgs, maintaxa = maintaxa, label = label,
            analysis = analysis, outcome = outcome,
            cohort = cohort, model_name = model_name, 
            n_obs = n_obs, est = est,
            se = se,ci_low = ci_low, ci_high = ci_high, p = p, 
            p_fdr_per_outcome = p_fdr_per_outcome)


merge_all_with_sig_subset <- merge_all_with_sig_subset %>% 
  mutate(phenotype = ifelse(model_name == 'full_stratM2_WHR_adjBMI_females', 'Female', 
                            ifelse(model_name == 'full_stratM2_WHR_adjBMI_males', 'Male', 
                                   ifelse(model_name == 'full_M2_WHR_adjBMI', 'All',
                                          'Interaction'))))
```

### re-structure the data to plot in the heatmap
```{r chunk created by Mariam 10 feb 2023}
transform_to_table <- merge_all_with_sig_subset %>% spread(phenotype, p_fdr_per_outcome)

transform_to_table<- transform_to_table %>%
  group_by(mgs) %>%
  summarise( label = label,
             All = na.omit(All),
             Female = na.omit(Female),
             Male = na.omit(Male),
             Interaction = na.omit(Interaction))


transform_to_table <- unique(transform_to_table) 

#use this table for the heatmap
transform_to_table <- transform_to_table %>% 
  mutate(All = signif(All,2),
         Female = signif(Female,2),
         Male = signif(Male,2),
         Interaction = signif(Interaction,2))

```

```{r}
merge_all_with_sig_subset <- merge_all_with_sig_subset[merge_all_with_sig_subset$phenotype != 'Interaction',]
merge_all_with_sig_subset <- merge_all_with_sig_subset[order(merge_all_with_sig_subset$mgs),]
export(merge_all_with_sig_subset, '/Volumes/research/metabogut/BMI_2021/summaryresults/MGS/whr_sigAll_females_males.xlsx')
```
